processor/awscontainerinsightprocessor/README.md

```diff
@@ -0,0 +1,445 @@
+# AWS Container Insights Processor
+
+## Overview
+AWS Container Insights Processor (`awscontainerinsightprocessor`) is an AWS specific processor that supports [CloudWatch Container Insights]((https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/ContainerInsights.html)). It decorates the metrics emitted by `awscontainerinsightreceiver` to support the same CloudWatch Container Insights experience. Unlike [kubeletstatsreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/kubeletstatsreceiver) and [k8sprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/k8sprocessor) which can work indepedently, `awscontainerinsightprocessor` is supposed to be used togther with `awscontainerinsightreceiver` and it expects certain metrics and resource attributes to be present in order to do the decoration (see [readme for `awscontainerinsightreceiver`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/awscontainerinsightreceiver/README.md)). 
```

Comment:
I am still not able to understand when to use `awscontainerinsightprocessor` and when to use `k8sprocessor`? Why do we need to have special one for awscontainerinsight? Can we make these 2 components (if we need to have separate components) work similar with the k8s once?

Reply:
I think the short answer is: don't use awscontainerinsightprocessor unless you want to use AWS Container Insights. 
As there are a lot of specific logic for AWS container insights in the code, for example, we are checking the type of metrics (e.g. node-level, pod-level, container-level) and do the processing differently. It would be difficult to design it as a generic processor that works for everyone. This processor itself doesn't call k8s api server and the relevant code is included in another PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3367. There is some discussion there about what to do in the future as there are already duplicated code for k8s in upstream.

Reply:
I'd ask the question from a slightly different perspective.  If this processor is intended to only be used in conjunction with a specific receiver, why isn't it simply part of that receiver?

Reply:
There is going to be another container insights receiver for EKS fargate that wants to use this processor to do metric decoration. That's why we want to separate the code for reuse. I think @hossain-rayhan knows more details about the receiver for EKS fargate.

Reply:
For EKS Fargate container insights support, we use the existing prometheus receiver. Using the prometheus receiver, we get resource usage metrics from cAdvisor metrics endpoint. From cAdvisor metrics endpoint, we don't get all the metrics we need to support EKS Fargate container insights. For example, we don't have a way to get the following metrics-
`pod_number_of_running_containers`
`container_number_of_restarts`
`pod_memory_request`

This `containerinsightsprocessor` generates these metrics and to support Fargate use case, we plan to use this common processor. 

Also, we need to calculate `containner.cpu.utilized` and `pod.cpu.utilized` metrics. To calculate these cpu utilized metrics, we need to store the previous record/datapoint. As this behavior is very specific to Amazon CloudWatch, I and Jaana discussed and planned to put it in this container inishgts processor. 

In short, to support CloudWatch container insights specific requirements, this custom processor is helpful. Though it requires a lot of changes to support Fargate cases. I think it only considers EKS EC2 now and many things are hardcoded. I can share details after having a deeper look.

Reply:
As a rule of thumb if this "scrapes" and "generates" metrics read from a target then is a receiver not a processor. Processor should only process already collected data, it may generate metrics but only from other already scraped metrics. Based on the comment it seems that you do read external targets from this processor.